OpenRA.Mods.Common/Traits/BotModules/McvExpansionManagerBotModule.cs

```diff
@@ -0,0 +1,767 @@
+
+		/* those are CheckBase mode options */
+		[Desc("Minimum distance in cells from center of the base expansion when checking for MCV deployment location.")]
+		public readonly int CBmodeMinDeployRadius = 2;
+
+		[Desc("Maximum distance in cells from center of the base expansion when checking for MCV deployment location.")]
+		public readonly int CBmodeMaxDeployRadius = 20;
+
+		public override object Create(ActorInitializer init) { return new McvExpansionManagerBotModule(init.Self, this); }
+	}
+
+	public class McvExpansionManagerBotModule : ConditionalTrait<McvExpansionManagerBotModuleInfo>, IBotTick, IBotRespondToAttack
+	{
+		const int PositiveMaxFailedAttempts = 3;
+		const int NegativeMaxFailedAttempts = 1;
+
+		readonly World world;
+		readonly Player player;
+		readonly ActorIndex.OwnerAndNamesAndTrait<TransformsInfo> mcvs;
+		readonly ActorIndex.OwnerAndNamesAndTrait<BuildingInfo> constructionYards;
+		readonly ActorIndex.OwnerAndNamesAndTrait<BuildingInfo> mcvFactories;
+		readonly int resourceIndiceStride;
+
+		IBotPositionsUpdated[] notifyPositionsUpdated;
+		IBotRequestUnitProduction[] requestUnitProduction;
+
+		int scanInterval;
+		int buildMCVInterval;
+		int moveConyardInterval;
+		int updateResourceMapInterval;
+		bool firstTick = true;
+		bool firstUndeploy = true;
+		bool allowfallback;
+
+		BotMcvExpansionMode mcvExpansionMode;
+		int mcvDeploymentMinDeployRadius;
+		int mcvDeploymentMaxDeployRadius;
+		int mcvDeploymentTryMaintainRange;
+
+		int maxFailedAttempts = 3;
+		int failedAttempts;
+		CPos? lastFailedExpandSpot;
+
+		int attackrespondcooldown = 20;
+
+		int pathDistanceSquareFactor;
+
+		BitArray resourceTypeIndices = null;
+		(CPos Center, int Value)[] resourceMapIndices = null;
+		int resourceMapIndicesColumnCount;
+		int resourceMapIndicesRowCount;
+		int updateResourceMapIndex;
+
+		public McvExpansionManagerBotModule(Actor self, McvExpansionManagerBotModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+			mcvs = new ActorIndex.OwnerAndNamesAndTrait<TransformsInfo>(world, info.McvTypes, player);
+			constructionYards = new ActorIndex.OwnerAndNamesAndTrait<BuildingInfo>(world, info.ConstructionYardTypes, player);
+			mcvFactories = new ActorIndex.OwnerAndNamesAndTrait<BuildingInfo>(world, info.McvFactoryTypes, player);
+			resourceIndiceStride = info.ResourceMapStride;
+		}
+
+		protected override void Created(Actor self)
+		{
+			// Special case handling is required for the Player actor.
+			// Created is called before Player.PlayerActor is assigned,
+			// so we must query player traits from self, which refers
+			// for bot modules always to the Player actor.
+			notifyPositionsUpdated = self.TraitsImplementing<IBotPositionsUpdated>().ToArray();
+			requestUnitProduction = self.TraitsImplementing<IBotRequestUnitProduction>().ToArray();
+			moveConyardInterval = Info.FirstMoveConyardTicks.RandomOrDefault(world.LocalRandom);
+			if (moveConyardInterval == -1)
```

Comment:
```suggestion
			if (moveConyardInterval < 0))
```

More sane interpretation

Reply:
fixed